File: chapter-13/compute/compute.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from functools import reduce


def is_even(n):
    return n % 2 == 0


def repeat(n, times):
    return [n] * times


def concat(list_a, list_b):
    return [*list_a, *list_b]


def computed():
    numbers = range(1, 8)
    evens = filter(is_even, numbers)
    repeats = map(lambda n: repeat(n, n), evens)
    values = reduce(concat, repeats, [])
    return values


print(computed())
# > python compute.py
# [2, 2, 4, 4, 4, 4, 6, 6, 6, 6, 6, 6]

File: chapter-13/compute/compute.js
function range(m, n) {
  return [...Array(n - m).keys()]
    .map(x => x + m);
}

function isEven(n) {
  return n % 2 === 0;
}

function repeat(n, times) {
  return Array(times).fill(n);
}

function concat(list_a, list_b) {
  return [...list_a, ...list_b]
}

function computed() {
  return range(1, 8)
    .filter(isEven)
    .flatMap(n => repeat(n, n));
}

console.log(computed())
